### ISSUE TITLE
Function hole leads to internal error

### ISSUE THREAD

**Issue by ysangkok** [idris-lang/Idris2#1216]
# Steps to Reproduce
```
main : IO ()
main = let
  x : String
  (?h x) = Just "hi!"
  in
  print x
```

# Expected Behavior
Should report that the hole `?h` is a constructor of type `String -> Maybe String`.

# Observed Behavior
`Error: INTERNAL ERROR: ?h is not a function application`

**Comment by andrevidela** [comment]
It took me a moment to understand what was going on here. I think Idris is parsing this as a clause with `?h x` as the left hand side and `Just "hi!"` as the right hand side, the left side should be a pattern and indeed something like `Just x` would be a valid pattern. However, holes cannot appear in patterns and so Idris aborts with an ugly error message because it's been given something that shouldn't have been possible in the first place.

**Comment by gallais** [comment]
> holes cannot appear in patterns

That was my first reaction too but then it got me wondering: is there
a deep reason for this restriction?

**Comment by gallais** [comment]
The fix should be fairly similar to that of #1035 so I'll mark this as a good first issue.

**Comment by gallais** [comment]
This was fixed in #1816.
